dbms/src/Storages/KVStore/MultiRaft/RegionSerde.cpp

```diff
@@ -0,0 +1,279 @@
+    // Increase this when persist with a new extension type.
+    UInt32 expected_extension_count = 0;
+    using bundle_type = std::pair<int, int>;
+    fiu_do_on(FailPoints::force_region_persist_version, {
+        if (auto v = FailPointHelper::getFailPointVal(FailPoints::force_region_persist_version); v)
+        {
+            std::tie(binary_version, expected_extension_count) = std::any_cast<bundle_type>(v.value());
+            LOG_WARNING(
+                Logger::get(),
+                "Failpoint force_region_persist_version set region binary version, value={}, "
+                "expected_extension_count={}",
+                binary_version,
+                expected_extension_count);
+        }
+    });
+    size_t total_size = writeBinary2(binary_version, buf);
+    UInt64 applied_index = -1;
+
+    {
+        std::shared_lock<std::shared_mutex> lock(mutex);
+
+        // Serialize meta
+        const auto [meta_size, index] = meta.serialize(buf);
+        total_size += meta_size;
+        applied_index = index;
+
+        // Try serialize extra flags
+        if (binary_version >= 2)
+        {
+            static_assert(sizeof(eager_truncated_index) == sizeof(UInt64));
+            UInt32 flags = RegionPersistFormat::HAS_EAGER_TRUNCATE_INDEX;
+            flags |= (expected_extension_count << 1);
+            total_size += writeBinary2(flags, buf);
+            total_size += writeBinary2(eager_truncated_index, buf);
+        }
+
+        UInt32 actual_extension_count = 0;
+        fiu_do_on(FailPoints::force_region_persist_extension_field, {
+            if (auto v = FailPointHelper::getFailPointVal(FailPoints::force_region_persist_extension_field); v)
+            {
+                auto value = std::any_cast<int>(v.value());
+                if (value & 1)
+                {
+                    std::string s = "abcd";
+                    total_size += writePersistExtension(
+                        actual_extension_count,
+                        buf,
+                        magic_enum::enum_underlying(RegionPersistExtension::ReservedForTest),
+                        s.data(),
+                        s.size());
+                }
+                if (value & 2)
+                {
+                    std::string s = "kkk";
+                    total_size += writePersistExtension(
+                        actual_extension_count,
+                        buf,
+                        UNUSED_EXTENSION_NUMBER_FOR_TEST,
+                        s.data(),
+                        s.size());
+                }
+                if (value & 4)
+                {
+                    std::string s = "zzz";
+                    total_size += writePersistExtension(
+                        actual_extension_count,
+                        buf,
+                        UNUSED_EXTENSION_NUMBER_FOR_TEST,
+                        s.data(),
+                        s.size());
+                }
+            }
+        });
```

Comment:
seems too hacky...

Reply:
I found no better idea

Reply:
separate `serialize()` into 

```
serializeMeta()
serializeExtraFlag()
serializeExtension()
serializeData()
```

and then we can mock serializeExtension just in ut?

Reply:
I can introduce a function for `serializeExtension`. However, the test for Region::serialize and Region::deserialize should be end-to-end. Because what I want to solve is compatibility between former different versions. I am can't split here, and test the mocked `serializeExtension`.

Reply:
ok, I have a better idea

```cpp
template<UInt32 version>
std::tuple<size_t, UInt64> Region::serialize(WriteBuffer & buf) const
{
    .......
    serializeExtension<version>(buf)
    .......
}

template<UInt32 version>
size_t serializeExtension(WriteBuffer & buf) const
{
    if constexpr (version >= 1)
    {
        // serialize Extension 1
        // you can mock some features here, just remove it and change to the real code when ungrade version.
    }
    if constexpr (version >= 2)
    {
        // serialize Extension 2
    }
    ......
}
```